skbio/stats/composition.py

```diff
@@ -807,6 +802,232 @@ def centralize(mat):
     return perturb_inv(mat, cen)
 
 
+@experimental(as_of="0.5.8")
+def balance_basis(tree_node: skbio.TreeNode):
+    r""" Determines the basis based on bifurcating tree.
+
+    This is commonly referred to as sequential binary partition [1]_.
+    Given a binary tree relating a list of features, this module can
+    be used to calculate an orthonormal basis, which is used to
+    calculate the ilr transform.
+
+    Parameters
+    ----------
+    treenode : skbio.TreeNode
+        Input bifurcating tree.  Must be strictly bifurcating
+        (i.e. every internal node needs to have exactly 2 children).
+
+    Returns
+    -------
+    basis : np.array
+        Returns a set of orthonormal bases in the Aitchison simplex
+        corresponding to the tree. The order of the
+        basis is index by the level order of the internal nodes.
+    nodes : list, str
+        List of tree nodes indicating the ordering in the basis.
+
+    Raises
+    ------
+    ValueError
+        The tree doesn't contain two branches.
+
+    Examples
+    --------
+    >>> from skbio.stats.composition import balance_basis
+    >>> from skbio import TreeNode
+    >>> tree = u"((b,c)a, d)root;"
+    >>> t = TreeNode.read([tree])
+    >>> basis, nodes = balance_basis(t)
+    >>> basis
+    array([[-0.40824829, -0.40824829,  0.81649658],
+           [-0.70710678,  0.70710678,  0.        ]])
+
+    Notes
+    -----
+    The tree must be strictly bifurcating, meaning that
+    every internal node has exactly 2 children.
+
+    See Also
+    --------
+    skbio.stats.composition.ilr
+
+    References
+    ----------
+    .. [1] J.J. Egozcue and V. Pawlowsky-Glahn "Exploring Compositional Data
+        with the CoDa-Dendrogram" (2011)
+
+    """
+    counts, n_tips = _count_matrix(tree_node)
+    counts = dict([(x, counts[x])
+                   for x in counts.keys() if not x.is_tip()])
+    nds = counts.keys()
+    r = np.array([counts[n]['r'] for n in nds])
+    s = np.array([counts[n]['l'] for n in nds])
+    k = np.array([counts[n]['k'] for n in nds])
+    t = np.array([counts[n]['t'] for n in nds])
+
+    a = np.sqrt(s / (r * (r + s)))
+    b = -1 * np.sqrt(r / (s * (r + s)))
+
+    basis = np.zeros((n_tips - 1, n_tips))
```

Comment:
That's going to be a big matrix...

Reply:
yea ... that's why a sparse matrix is preferable